### PR TITLE
Deploy stack services one-by-one

### PR DESCRIPTION
**Pull request by jakolehm**
Fixes stack deployments so that stack services are deployed one-by-one and with correct order.

### CODE REVIEW

**Comment by jakolehm** [comment]
@jnummelin ping

File: server/app/jobs/stack_deploy_worker.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    remove_services(stack, stack_rev)
+    services = sort_services(stack.grid_services.to_a)
+    services.each do |service|
+      deploy_service(service, stack_deploy)
+    end
+    stack_deploy.success!
+
+    stack_deploy
+  rescue
+    stack_deploy.error!
```
Comment (jnummelin) [inline server/app/jobs/stack_deploy_worker.rb]: Could we somehow get the possibile service deployment error messages appended here?

File: server/app/jobs/stack_deploy_worker.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  # @param [StackDeploy] stack_deploy
+  # @param [StackRevision] stack_rev
+  def deploy_stack(stack_deploy, stack_rev)
+    stack = stack_deploy.stack
+    stack_deploy.ongoing!
+
+    remove_services(stack, stack_rev)
+    services = sort_services(stack.grid_services.to_a)
+    services.each do |service|
+      deploy_service(service, stack_deploy)
```
Comment (jnummelin) [inline server/app/jobs/stack_deploy_worker.rb]: Just thinking, but should the failure in one service cause others not to be deployed?

**Comment by jakolehm** [comment]
@SpComb also ping :)

File: server/app/jobs/stack_deploy_worker.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            info "waiting for #{service.to_path} deploy to finish..."
+            sleep 1
+            if service_deploy.reload && (service_deploy.success? || service_deploy.error?)
+              finished = true
+            end
+          end
+        end
+      rescue => exc
+        error "#{exc.class.name}: #{exc.message}"
+      end
```
Comment (jnummelin) [inline server/app/jobs/stack_deploy_worker.rb]: Maybe return the deployment object to be able to collect failure reasons at stack level.

File: server/app/models/stack_deploy.rb [hunk 4]
```diff
@@ -1,9 +1,12 @@
 class StackDeploy
   include Mongoid::Document
   include Mongoid::Timestamps
+  include Mongoid::Enum
 
   index({ stack_id: 1 })
 
+  enum :deploy_state, [:created, :ongoing, :success, :error], default: :created
+
```
Comment (jnummelin) [inline server/app/models/stack_deploy.rb]: Add also a `reason`field as the CLI is expecting that on `wait_deploy_to_finish`

**Comment by jakolehm** [comment]
Stack deploy will now halt when first service errors. And reason is displayed to the user:

```
$ kontena stack install --deploy test.yml
 [done] Creating stack test      
 [fail] Deploying stack test     
        Cannot find applicable node for service instance test/test/nginx-1
```

**Pull request merged**